core/src/runtime/mod.rs

```diff
@@ -1255,6 +1287,73 @@ impl<'a> Runtime<'a> {
     }
 }
 
+pub trait LookupIdSampler {
+    fn sample(&mut self, num_lookup_ids: usize) -> &[u128];
+}
+
+impl<'a> LookupIdSampler for Runtime<'a> {
+    fn sample(&mut self, num_lookup_ids: usize) -> &[u128] {
+        if self.preallocated_random_nums.is_empty()
+            || (self.preallocated_random_nums_idx + num_lookup_ids) >= NUM_PREALLOCATED_RANDOM_NUMS
+        {
+            let chunk_size = std::cmp::max(NUM_PREALLOCATED_RANDOM_NUMS / num_cpus::get(), 1);
+            self.preallocated_random_nums = vec![0; NUM_PREALLOCATED_RANDOM_NUMS];
+
+            self.preallocated_random_nums
+                .chunks_mut(chunk_size)
+                .par_bridge()
+                .for_each(|chunk| {
+                    let mut rng = rand::thread_rng();
+                    chunk.iter_mut().for_each(|num| {
+                        *num = rng.gen();
+                    });
+                });
+
+            self.preallocated_random_nums_idx = 0;
+        }
+
+        let rnd_nums = self.preallocated_random_nums
```

Comment:
instead of doing this you should just use VecDeque and then just push_back and pop_front.

Reply:
I was considering using a VecDeque, but I don't think it supports multi threaded inserts.  That's one of the main performance gains from the PR.